beacon_node/execution_layer/src/engine_api/http.rs

```diff
@@ -625,40 +663,121 @@ impl HttpJsonRpc {
             .header(CONTENT_TYPE, "application/json")
             .json(&body);
 
-        // Generate and add a jwt token to the header if auth is defined.
-        if let Some(auth) = &self.auth {
-            request = request.bearer_auth(auth.generate_token()?);
-        };
-
-        let body: JsonResponseBody = request.send().await?.error_for_status()?.json().await?;
-
-        match (body.result, body.error) {
-            (result, None) => serde_json::from_value(result).map_err(Into::into),
-            (_, Some(error)) => {
-                if error.message.contains(EIP155_ERROR_STR) {
-                    Err(Error::Eip155Failure)
-                } else {
-                    Err(Error::ServerMessage {
-                        code: error.code,
-                        message: error.message,
-                    })
+            // Generate and add a jwt token to the header if auth is defined.
+            if let Some(auth) = &self.auth {
+                request = request.bearer_auth(auth.generate_token()?);
+            };
+
+            let body: JsonResponseBody = request.send().await?.error_for_status()?.json().await?;
+
+            match (body.result, body.error) {
+                (result, None) => serde_json::from_value(result).map_err(Into::into),
+                (_, Some(error)) => {
+                    if error.message.contains(EIP155_ERROR_STR) {
+                        Err(Error::Eip155Failure)
+                    } else {
+                        Err(Error::ServerMessage {
+                            code: error.code,
+                            message: error.message,
+                        })
+                    }
+                }
+            }
+        }
+        else {
+
+            let mut wspackage = self.wspackage.lock().await;
+            wspackage.id += 1;
+
+            if wspackage.status == 0 {
+                if let Some(auth) = &self.auth {
+                    let wsrouter = WsRouter::new_with_jwt(self.url.full.to_string(), auth.generate_token()?).await.map_err(|e| Error::WebsocketError(e.to_string()))?;
+                    wspackage.wsrouter = Some(wsrouter);
+                    wspackage.status = 1;
+                }
+                else {
+                    let wsrouter = WsRouter::new(self.url.full.to_string()).await.map_err(|e| Error::WebsocketError(e.to_string()))?;
+                    wspackage.wsrouter = Some(wsrouter);
+                    wspackage.status = 1;
                 }
             }
+
+            let body = JsonRequestBody {
+                jsonrpc: JSONRPC_VERSION,
+                method,
+                params,
+                id: json!(&wspackage.id),
+            };
+
+            let payload = serde_json::to_string(&body)?;
+            // map Timeout error to our own
+            // map Other to our own
+            let response = wspackage.wsrouter.as_ref().unwrap().make_request_timeout(payload.clone(), wspackage.id, timeout).await;
```

Comment:
I think if you used `ajsonrpc`'s `.send()` here you could drop the `wspackage` lock while waiting for the response. This would allow requests to run in parallel.